lib/common/xml.c

```diff
@@ -261,30 +256,54 @@ pcmkDeregisterNode(xmlNodePtr node)
        -> xsltFreeTransformContext -> xsltFreeRVTs -> xmlFreeDoc)
        onto result tree fragments, represented as standalone documents
        with otherwise infeasible space-prefixed name (xsltInternals.h:
-       XSLT_MARK_RES_TREE_FRAG) and carrying it's own load at _private
-       field -- later assert on the XML_PRIVATE_MAGIC would explode */
-    if (node->type != XML_DOCUMENT_NODE || node->name == NULL
-            || node->name[0] != ' ') {
-        __xml_private_free(node->_private);
+       XSLT_MARK_RES_TREE_FRAG) and carrying its own payload at _private
+       field -- later assert on the XML_*_PRIVATE_MAGIC would explode */
+    if (node->name == NULL || node->name[0] != ' ') {
+        if (node->_private) {
+            if (node->type == XML_DOCUMENT_NODE) {
+                __xml_doc_private_clean(node->_private);
+            } else {
+                CRM_ASSERT(((xml_node_private_t *) node->_private)->check
+                               == XML_NODE_PRIVATE_MAGIC);
+                /* nothing dynamically allocated nested */
+            }
+            free(node->_private);
+            node->_private = NULL;
+        }
     }
 }
 
 static void
 pcmkRegisterNode(xmlNodePtr node)
 {
-    xml_private_t *p = NULL;
-
-    switch(node->type) {
+    switch (node->type) {
+        case XML_DOCUMENT_NODE: {
+            xml_doc_private_t *docpriv = NULL;
+            docpriv = calloc(1, sizeof(xml_doc_private_t));
+            docpriv->check = XML_DOC_PRIVATE_MAGIC;
+            /* Flags will be reset if necessary when tracking is enabled */
+            docpriv->flags |= (xpf_dirty|xpf_created);
+            node->_private = docpriv;
+            break;
+        }
         case XML_ELEMENT_NODE:
-        case XML_DOCUMENT_NODE:
         case XML_ATTRIBUTE_NODE:
-        case XML_COMMENT_NODE:
-            p = calloc(1, sizeof(xml_private_t));
-            p->check = XML_PRIVATE_MAGIC;
+        case XML_COMMENT_NODE: {
+            xml_node_private_t *nodepriv = NULL;
+            nodepriv = calloc(1, sizeof(xml_node_private_t));
+            nodepriv->check = XML_NODE_PRIVATE_MAGIC;
             /* Flags will be reset if necessary when tracking is enabled */
-            p->flags |= (xpf_dirty|xpf_created);
-            node->_private = p;
+            nodepriv->flags |= (xpf_dirty|xpf_created);
+            node->_private = nodepriv;
+            if (nodepriv && pcmk__tracking_xml_changes(node, FALSE)) {
+                /* XML_ELEMENT_NODE doesn't get picked up here, node->doc is
+                 * not hooked up at the point we are called
+                 */
+                pcmk__set_xml_flag(node, xpf_dirty);
+                __xml_node_dirty(node);
```

Comment:
__xml_node_dirty(node) calls pcmk__set_xml_flag(node, xpf_dirty), so I don't think that explicit call was ever necessary